sway/desktop/output.c

```diff
@@ -755,8 +755,94 @@ static void render_container(struct sway_output *output,
 		render_container_tabbed(output, damage, con, parent_focused);
 		break;
 	case L_FLOATING:
-		// TODO
-		break;
+		sway_assert(false, "Didn't expect to see floating here");
+	}
+}
+
+static bool floater_intersects_output(struct sway_container *floater,
+		struct sway_container *output) {
+	struct wlr_box box = {
+		.x = floater->x,
+		.y = floater->y,
+		.width = floater->width,
+		.height = floater->height,
+	};
+	return wlr_output_layout_intersects(root_container.sway_root->output_layout,
+			output->sway_output->wlr_output, &box);
+}
+
+static void container_translate(struct sway_container *con, int x, int y) {
+	con->x += x;
+	con->y += y;
+	if (con->type == C_VIEW) {
+		con->sway_view->x += x;
+		con->sway_view->y += y;
+	} else {
+		for (int i = 0; i < con->children->length; ++i) {
+			struct sway_container *child = con->children->items[i];
+			container_translate(child, x, y);
+		}
+	}
+}
+
+static void render_floating_container(struct sway_output *soutput,
+		pixman_region32_t *damage, struct sway_container *con) {
+	// We need to translate the floating container's coordinates from layout
+	// coordinates into output-local coordinates. This needs to happen for all
+	// children of the floating container too.
+	struct sway_container *output = container_parent(con, C_OUTPUT);
+	container_translate(con, -output->x, -output->y);
```

Comment:
Hmm, I'm really not a fan of this. Find another way.

Reply:
Yeah, I'm not a fan either. The options I can think of for this are:

1) Adjust the container's position temporarily as it currently does.
2) Have file-scoped `translate_x` and `translate_y` variables, which are 0 if the container being rendered is tiled, and the inverse of the output's position if floating. The rendering functions would add the translation to the thing being rendered, so it's basically forced into output-level scope.
3) Same as above, but using a `wlr_matrix` instead of the translate variables.
4) Use layout-local coordinates for all swayc positions, then subtract the output's position when adding damage and rendering anything.

Options 2 and 3 are pretty complicated. To damage or render any view, you have to figure out if the view is itself floating or a child of floating container. If it is, its coordinates are layout-local and need to be translated to output-local coordinates.

I think the best option is option 4 - use layout-local coordinates for all swaycs. This would also resolve the "these coords might be layout-local or output-local" issue that you pointed out in https://github.com/swaywm/sway/pull/2027#pullrequestreview-123501607.

Reply:
I went and implemented option 4. That particular commit was +71, -130. It has simplified other areas of the codebase and appears to have fixed my damage problems too.